File: crates/swiftlet/examples/calculate.rs
use std::sync::Arc;
use swiftlet::preclude::*;

fn calculate(ast: &Ast) -> i32 {
    match ast {
        Ast::Token(token) => token.word().parse::<i32>().unwrap(),
        Ast::Tree(tree, children) => match tree.as_str() {
            "start" | "expr" => calculate(&children[0]),
            "add" => calculate(&children[0]) + calculate(&children[2]),
            "sub" => calculate(&children[0]) - calculate(&children[2]),
            _ => {
                panic!("Invalid tree: {}", tree);
            }
        },
    }
}

fn main() {
    let grammar = r#"
        start: expr
        expr: expr "+" INT -> add
            | expr "-" INT -> sub
            | INT
        %import (WS, INT)
        %ignore WS
        "#;

    let conf = Arc::new(ParserConfig {
        algorithm: Algorithm::Earley,
        debug: true,
        ..Default::default()
    });
    let parser = Swiftlet::from_str(grammar)
        .map(|grammar| grammar.parser(conf))
        .expect("failed to get parser");
    let text = "10 - 2 + 5 - 2";

    match parser.parse(text) {
        Ok(tree) => {
            println!("AST: ");
            tree.pretty_print();
            println!("Total: {}", calculate(&tree));
        }
        Err(e) => {
            println!("Error: {}", e);
        }
    }
}

File: pyswiftlet/examples/calculate.py
from swiftlet import Swiftlet, Transformer

class Calculate(Transformer):
    def start(self, child):
        return child[0]

    def expr(self, child):
        return child[0]

    def add(self, child):
        return child[0] + child[2]

    def sub(self, child):
        return child[0] - child[2]

    def mul(self, child):
        return child[0] * child[2]

    def div(self, child):
        return child[0] / child[1]

    def terms(self, child):
        return child[0]

def main():
    grammar = """
    start: expr
    expr: expr "+" terms -> add
        | expr "-" terms -> sub
        | terms
    terms: terms "*" INT -> mul
        | terms "/" INT -> div
        | INT
    %import (WS, INT)
    %ignore WS
    """

    text = "12 + 10 - 8 * 2 + 4"
    parser = Swiftlet(grammar)
    parsed_text = parser.parse(text)

    calculate = Calculate()
    result = calculate(parsed_text)

    parsed_text.pretty_print()
    print("\nResult: ", result)


if __name__ == "__main__":
    main()